src/circuit.rs

```diff
@@ -50,13 +270,401 @@ impl plonk::Circuit<pallas::Base> for Circuit {
+                                    -signed_value
+                                } else {
+                                    signed_value
+                                };
+                                assert!(magnitude < pallas::Scalar::from_u128(1 << 64));
+                                pallas::Base::from_bytes(&magnitude.to_bytes()).unwrap()
+                            };
+                            let sign = if is_negative {
+                                -pallas::Base::one()
+                            } else {
+                                pallas::Base::one()
+                            };
+                            (magnitude, sign)
+                        });
+                let magnitude = self.load_private(
+                    layouter.namespace(|| "v_net magnitude"),
+                    config.advices[9],
+                    magnitude_sign.map(|m_s| m_s.0),
+                )?;
+                let sign = self.load_private(
+                    layouter.namespace(|| "v_net sign"),
+                    config.advices[9],
+                    magnitude_sign.map(|m_s| m_s.1),
+                )?;
+                (magnitude, sign)
+            };
+
+            // Constrain v_old - v_new = magnitude * sign
+            layouter.assign_region(
+                || "v_old - v_new = magnitude * sign",
+                |mut region| {
+                    copy(&mut region, || "v_old", config.advices[0], 0, &v_old)?;
+                    copy(&mut region, || "v_new", config.advices[1], 0, &v_new)?;
+                    let (magnitude, sign) = v_net;
+                    copy(
+                        &mut region,
+                        || "v_net magnitude",
+                        config.advices[2],
+                        0,
+                        &magnitude,
+                    )?;
+                    copy(&mut region, || "v_net sign", config.advices[3], 0, &sign)?;
+
+                    config.q_v_net.enable(&mut region, 0)
+                },
+            )?;
+
+            // commitment = [v_net] ValueCommitV
+            let (commitment, _) = {
+                let value_commit_v = ValueCommitV::get();
+                let value_commit_v = FixedPointShort::from_inner(ecc_chip.clone(), value_commit_v);
+                value_commit_v.mul(layouter.namespace(|| "[v_net] ValueCommitV"), v_net)?
+            };
+
+            // blind = [rcv] ValueCommitR
+            let (blind, _) = {
+                let rcv = self.rcv.as_ref().map(|rcv| **rcv);
+                let value_commit_r = OrchardFixedBasesFull::ValueCommitR;
+                let value_commit_r = FixedPoint::from_inner(ecc_chip.clone(), value_commit_r);
+
+                // [rcv] ValueCommitR
+                value_commit_r.mul(layouter.namespace(|| "[rcv] ValueCommitR"), rcv)?
+            };
+
+            // [v_net] ValueCommitV + [rcv] ValueCommitR
+            commitment.add(layouter.namespace(|| "cv_net"), &blind)?
```

Comment:
Why do we do this at the Action circuit level, rather than having an abstraction for homomorphic Pedersen commitments?

Reply:
Opened https://github.com/zcash/orchard/issues/158 for this.